cmake/ext_project_autotools.cmake

```diff
@@ -0,0 +1,118 @@
+include(utils)
+
+# Build a third-party project that uses autotools to generate build system.
+# Basically it generates configure script, run configure and then make and
+# rerun them according to their dependants in case of incremental build.
+#
+# DIR - relative path to project directory in main project tree.
+# CONFIGURE - options for configure script.
+# BYPRODUCTS - products caller targets depend upon.
+#
+# Note that incremental rebuild in case autotools files or config files are
+# changed is available only if CMake version is at least 3.12. Otherwise
+# rebuild will not be triggered. Rebuild is only required for developers
+# which presumably has newer CMake versions.
+function(ext_project_autotools name)
```

Comment:
(A chart to ease discussion.)

```mermaid
flowchart TD
    cmd_source_update([git submodule update --recursive --init])
    cmd_autoreconf([autoreconf -i])
    cmd_configure([./configure <...>])
    cmd_config_status([./config.status])
    cmd_make([make <...>])

    files_am[*.am]
    files_in[*.in]
    files_configure[configure]
    files_makefile[Makefile]
    files_config_stamp[config.stamp.tarantool]
    files_byproducts[$byproducts]

    cmd_source_update --> files_am
    cmd_source_update --> files_in
    files_am --> cmd_autoreconf
    files_in --> cmd_config_status
    cmd_autoreconf --> files_configure
    cmd_autoreconf -.-> files_in
    files_configure --> cmd_configure
    cmd_configure --> files_makefile
    files_makefile --> cmd_config_status
    cmd_config_status --> files_config_stamp
    files_makefile --> cmd_make
    files_config_stamp --> cmd_make
    cmd_make -.-> files_byproducts
```

Reply:
Looks very similar to autotools graph at page [https://en.wikipedia.org/wiki/GNU_Autotools](https://en.wikipedia.org/wiki/GNU_Autotools) :)

Reply:
We discussed the `*.in` -> `./config.status` arrow with Nikolay.

There are `*.in` files, which are part of the repository, and there are `*.in` files that are generated by `autoreconf -i`. Changing of the former ones don't obligate us to pass over the whole autoreconf->configure->make pipeline. We can just call `./config.status` (which is effectively 'repeat ./configure with previously passed options') and `make`, it is faster.

Changing of generated `*.in` files means that `*.am` files were changed, so we'll pass over the whole autoreconf->configure->make pipeline anyway.

So the graph looks logical for me.